### PR TITLE
graph_trainer: coalesce chunked lm_head gradient sync

### PR DESCRIPTION
**Pull request by SherlockNoMad**
• ## Summary

  - Fix graph-trainer chunked loss issuing one lm_head reduce-scatter per chunk under simple FSDP.
  - Accumulate eight chunk gradients locally in FP32, then perform one final reduce-scatter.
  - Add explicit graph-trainer contracts: non-FSDP2, bias-free lm_head with one weight.
  - Add a separate local_map prototype that avoids hooks and .grad side effects.

  ## Verification

  - Traced graph: one all-gather, seven FP32 additions, one reduce-scatter.
  - Bitwise-equal loss, hidden gradient, and weight gradient against native FSDP2.
  - Traced replay is bitwise-equal to eager execution.
  - Six focused tests pass.

### CODE REVIEW

File: torchtitan/experiments/graph_trainer/chunked_loss.py [hunk 1]
```diff
@@ -36,33 +94,120 @@ def _gradient_backprop(
...
+        )
+        partial_grad = DTensor.from_local(
+            local_grad,
+            device_mesh=sharded_weight.device_mesh,
+            placements=(Partial(reduce_op="sum"),) * num_dp_axes + non_dp_placements,
+            run_check=False,
+            shape=sharded_weight.shape,
+            stride=sharded_weight.stride(),
+        )
+        sharded_weight.grad = partial_grad.redistribute(
```
Comment (weifengpy) [inline torchtitan/experiments/graph_trainer/chunked_loss.py:195]: is there a risk of overwriting previous `sharded_weight.grad`? 

```
lm_head_eager.zero_grad(set_to_none=True)

loss_fn = ChunkedLossWrapperWithParamGrads(
    ChunkedLossWrapperWithParamGrads.Config(num_chunks=num_chunks)
)
loss_fn.set_lm_head(lm_head_eager)

loss, _ = loss_fn(hidden_states, labels)
loss.backward()
grad_after_one = next(lm_head_eager.parameters()).grad.clone()

loss, _ = loss_fn(hidden_states, labels)
loss.backward()
grad_after_two = next(lm_head_eager.parameters()).grad
```

File: torchtitan/experiments/graph_trainer/chunked_loss.py [hunk 2]
```diff
@@ -72,53 +217,23 @@ def forward(
...
-            # param's AccumulateGrad firing on the grads we return below.
-            # If we restored here (synchronously, before returning), the
-            # first AccumulateGrad would see sync=True and try to
-            # reduce-scatter our already-sharded grad → wrong result.
-            lm_head = ctx.lm_head
-            torch.autograd.Variable._execution_engine.queue_callback(
-                lambda: lm_head.set_requires_gradient_sync(True, recurse=False)
-            )
+        weight_grad = saved[1]
         return (
```
Comment (weifengpy) [inline torchtitan/experiments/graph_trainer/chunked_loss.py:234]: codex warned that `(loss_scale * loss).backward()` might not work. i don't think titan has loss scaling,  thus this might be a nit

**Review by weifengpy — COMMENTED** [review]
finished review on what's minimal correctness change to make it landable
* .grad overwrite
* loss scaling (might be nit)

I never gated on simplefsdp change. @SherlockNoMad let me know which is your intention
* prototype something to show feasibility and let me take over to a landable state
* OR address above correctness issue directly, and i can do a second review
